### PR TITLE
feat: support server configuration as code (using immich-config.json)

### LINKED ISSUE #72: Config file support
Now that Immich supports YAML-formatted config files, support for those could be nicely integrated into the helm values

### PR DESCRIPTION
**Pull request by hiteshnayak305**
### Description of the change

Support configuration of immich settings using helm chart.

values.yaml example

```
...
immich:
  configuration: 
    trash:
      enabled: true
      days: 30
...
```

### Applicable issues

  - fixes #72 

### Open Point

No open points

### Use cases

- If `configuration:` or `configuration: {}` no changes to original behaviour or additional resources created and settings can be changed from UI.
- If valid values provided a ConfigMap `release-name-immich-config` will be created and mounted on `server` and `microservices` pod
  with new env variable `IMMICH_CONFIG_FILE: /config/immich-config.yaml` . No settings can be changed from UI after configuration.

### CODE REVIEW

File: charts/immich/values.yaml [hunk 1]
```diff
@@ -26,6 +26,35 @@ immich:
...
+  # can be provided as example below.
+  #
+  configuration: {}
+    # Example 
+    # trash:
+    #   enabled: false
+    #   days: 30
+    # storageTemplate:
+    #   enabled: true
+    #   template: "{{`{{`}}y{{`}}`}}-{{`{{`}}MM{{`}}`}}-{{`{{`}}dd{{`}}`}}/{{`{{`}}filename{{`}}`}}"
```
Comment (bo0tzz) [inline charts/immich/values.yaml:57]: Would it work to just escape the whole line, like 
```
template: "{{`{{y}}/{{y}}-{{MM}}-{{dd}}/{{filename}}`}}
```
?

File: charts/immich/values.yaml [hunk 2]
```diff
@@ -26,6 +26,35 @@ immich:
...
+  #   "storageTemplate": {
+  #     "enabled": true
+  #     "template": "{{y}}/{{y}}-{{MM}}-{{dd}}/{{filename}}"
+  #   }
+  # }
+  # can be provided as example below.
+  #
+  configuration: {}
+    # Example 
+    # trash:
```
Comment (bo0tzz) [inline charts/immich/values.yaml:52]:
```suggestion
  configuration: {}
    # trash:
```
It being commented out already indicates that it's an example ;)

File: charts/immich/values.yaml [hunk 3]
```diff
@@ -26,6 +26,35 @@ immich:
...
+  #     "enabled": true,
+  #     "days": 30
+  #   },
+  #   "storageTemplate": {
+  #     "enabled": true
+  #     "template": "{{y}}/{{y}}-{{MM}}-{{dd}}/{{filename}}"
+  #   }
+  # }
+  # can be provided as example below.
+  #
```
Comment (bo0tzz) [inline charts/immich/values.yaml:49]: I don't think we need this section

**Review by bo0tzz — CHANGES_REQUESTED** [review]
Awesome stuff, thank you! Some notes:

1. The config also needs to be mounted to the microservices pod. I think that means we'll need to create our own configmap template rather than using the common-library one (else we'd be creating two configmaps). Or maybe we can invoke the common-library configmap template directly?
2. Immich supports a yaml-formatted config file, which means we don't need to `toJson` it.

File: charts/immich/values.yaml (hunk 1 above)
Comment (hiteshnayak305) [reply charts/immich/values.yaml:57]: yes it works. thanks.

File: charts/immich/values.yaml (hunk 2 above)
Comment (hiteshnayak305) [reply charts/immich/values.yaml:52]: okay :+1:

File: charts/immich/values.yaml (hunk 3 above)
Comment (hiteshnayak305) [reply charts/immich/values.yaml:49]: :+1:

**Comment by hiteshnayak305** [comment]
For now using two separate config map for microservices and server.

1. Tried with own template for configmap but change in configuration is not rolling deployments on upgrade.
2.  I think same will occur when deploy configmap through library usage. But I don't have any refrence or docs.

**Comment by bo0tzz** [comment]
> Tried with own template for configmap but change in configuration is not rolling deployments on upgrade.

I think this issue happens because the Deployment isn't actually changing, and Kubernetes won't automatically reload configmaps. This would be fixed by generating a new configmap every time the config file changes (https://stackoverflow.com/a/40624029). We could do this for example by including the first section of the configuration's hash in the configmap name (also see https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments). 

> I think same will occur when deploy configmap through library usage. But I don't have any refrence or docs.

The library handles this by applying a trick like above: it adds the hash of the configmap as an annotation on the pod (see https://github.com/bjw-s/helm-charts/blob/923ef40a39520979c98f354ea23963ee54f54433/charts/library/common/templates/lib/metadata/_podAnnotations.tpl#L7-L15). However, I had a quick attempt at using the library's configmap and it seems more trouble than it's worth, and I also think it might not work nicely with the hashing that it does.

If you're up for it, let's try our own configmap template with the hash trick (or something like it).

**Comment by hiteshnayak305** [comment]
I looked into https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments , to achieve this we need to set annotation inside deployment spec which  is not exposed by library. Let me know if possible in any ways.

**Comment by bo0tzz** [comment]
I think it can be done with `podAnnotations` in the values. Another option might be to append (part of) the hash to the name of the configmap.

**Comment by hiteshnayak305** [comment]
Thanks. It worked. Using custom configMap also resolves condition for `{{`, no need to wrap in `` {{` ``.

**Review by bo0tzz — APPROVED** [review]
Amazing, thank you very much!

**Pull request merged**